### PR TITLE
[inductor] Enable padding fusion with torch.cat; Avoid duplicate computation in cat/pad when inputs have multiple consumers

### PR DESCRIPTION
**Pull request by tianrengao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #175729

## Summary

Fixes a performance issue where `torch.cat` and `F.pad` (constant_pad_nd)
generate redundant computation when their inputs have multiple consumers.

**Issues fixed:**
- pytorch/pytorch#125075 — Inductor cannot fuse cat with a pointwise (pr1)
- vllm-project/vllm#25477 — Fuse mul and pad in MoE models (pr2)

### Issue #125075 (cat + to_fp16)

```python
z = torch.cat([x, torch.zeros([6, 768])], dim=0)  # [30528, 768]
y = x.to(torch.float16)                            # [30522, 768]
```
x has two consumers (cat and to_fp16). Inductor's pointwise_cat heuristic inlines x into the cat kernel via tl.where masking, but x must also be realized separately for to_fp16, causing x to be loaded twice.

## Solution
Add any_input_has_multi_consumers() check in cat lowering: when any input FX node has multiple users, skip pointwise_cat and fall through to ConcatKernel. ConcatKernel uses NonOwningLayout to let the input write directly into the output's slice — zero copy, shared by all consumers.

### BEFORE (x loaded twice, 2 independent kernels):

```python
def call(self, args):
    arg0_1, = args
    # Kernel 1: pointwise_cat — iterates 30528*768, reads arg0_1, tl.where for padding
    buf0 = empty_strided_cuda((30528, 768), (768, 1), torch.float32)
    triton_poi_fused_cat_zeros_0.run(arg0_1, buf0, 23445504)

    # Kernel 2: to_fp16 — iterates 30522*768, reads arg0_1 AGAIN
    buf1 = empty_strided_cuda((30522, 768), (768, 1), torch.float16)
    triton_poi_fused__to_copy_1.run(arg0_1, buf1, 23440896)
    return (buf0, buf1)
```

Kernel 1 (pointwise_cat with tl.where):
```python
@triton.jit  # in_ptr0=*fp32, out_ptr0=*fp32
def triton_poi_fused_cat_zeros_0(in_ptr0, out_ptr0, xnumel, XBLOCK):
    xnumel = 23445504               # 30528 * 768
    x0 = (xindex % 768)
    x1 = xindex // 768
    tmp2 = x1 < 30522               # tl.where: in padding region?
    tmp5 = tl.load(in_ptr0 + ..., tmp2, other=0.0)  # masked load from arg0_1
    tl.store(out_ptr0 + (x0 + 768*x1), tmp5, ...)
```

Kernel 2 (separate to_fp16):
```python
@triton.jit  # in_ptr0=*fp32, out_ptr0=*fp16
def triton_poi_fused__to_copy_1(in_ptr0, out_ptr0, xnumel, XBLOCK):
    xnumel = 23440896               # 30522 * 768
    tmp0 = tl.load(in_ptr0 + (x0), xmask)  # reads arg0_1 AGAIN
    tmp1 = tmp0.to(tl.float32)
    tl.store(out_ptr0 + (x0), tmp1, xmask)  # *fp16 store → implicit downcast
```

**arg0_1 loaded twice. ~178MB memory bandwidth.**

### AFTER (x loaded once, fused kernel):

```python
def call(self, args):
    arg0_1, = args
    buf2 = empty_strided_cuda((30528, 768), (768, 1), torch.float32)     # cat output
    buf0 = reinterpret_tensor(buf2, (30522, 768), (768, 1), 0)           # alias, no alloc
    buf3 = empty_strided_cuda((30522, 768), (768, 1), torch.float16)     # fp16 output

    # Kernel 1: FUSED copy + to_fp16 — reads arg0_1 ONCE, writes 2 outputs
    triton_poi_fused__to_copy_cat_zeros_0.run(arg0_1, buf0, buf3, 23440896)

    buf1 = reinterpret_tensor(buf2, (6, 768), (768, 1), 23440896)       # pad alias
    # Kernel 2: fill zeros — only 4608 elements (0.02% of data)
    triton_poi_fused_zeros_1.run(buf1, 4608)
    return (buf2, buf3)
```

Fused kernel (1 load, 2 stores):
```python
@triton.jit  # in_ptr0=*fp32, out_ptr0=*fp32, out_ptr1=*fp16
def triton_poi_fused__to_copy_cat_zeros_0(in_ptr0, out_ptr0, out_ptr1, xnumel, XBLOCK):
    xnumel = 23440896               # 30522 * 768
    tmp0 = tl.load(in_ptr0 + (x0), xmask)   # load arg0_1 ONCE
    tmp1 = tmp0.to(tl.float32)               # noop (already fp32), triton optimizes away
    tl.store(out_ptr0 + (x0), tmp0, xmask)   # write to cat slice (fp32)
    tl.store(out_ptr1 + (x0), tmp1, xmask)   # write to fp16 output (implicit fp32→fp16)
```

**arg0_1 loaded once. ~89MB bandwidth saved. No tl.where branching.**

</div><h2 style="box-sizing: border-box; color: rgb(212, 212, 212); font-family: -apple-system, &quot;system-ui&quot;, sans-serif; font-weight: 400; line-height: 1.1; margin: 0px 0px 8px; font-size: 16.9px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Test Plan</h2><p style="box-sizing: border-box; margin: 0px 0px 8px; color: rgb(212, 212, 212); font-family: -apple-system, &quot;system-ui&quot;, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Added<span> </span><code style="font-family: Menlo, Monaco, &quot;Courier New&quot;, monospace; color: rgb(204, 204, 204); background-color: transparent; padding: 0px; border-radius: 4px; box-sizing: border-box; font-size: 1em; font-weight: 400; line-height: 1.357em;">TestCatMultiConsumer</code><span> </span>with 2 tests:</p><ul style="box-sizing: border-box; margin: 0px 0px 8px; color: rgb(212, 212, 212); font-family: -apple-system, &quot;system-ui&quot;, sans-serif; font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="box-sizing: border-box;"><code style="font-family: Menlo, Monaco, &quot;Courier New&quot;, monospace; color: rgb(204, 204, 204); background-color: transparent; padding: 0px; border-radius: 4px; box-sizing: border-box; font-size: 1em; font-weight: 400; line-height: 1.357em;">test_cat_to_fp16</code>: multi-consumer cat fuses correctly</li><li style="box-sizing: border-box;"><code style="font-family: Menlo, Monaco, &quot;Courier New&quot;, monospace; color: rgb(204, 204, 204); background-color: transparent; padding: 0px; border-radius: 4px; box-sizing: border-box; font-size: 1em; font-weight: 400; line-height: 1.357em;">test_single_consumer_cat_unchanged</code>: single-consumer behavior unchanged</li></ul>

Verified with repro scripts on NVIDIA B200

### other benchmarks:

Test pattern: z = cat([x, zeros([6,768])]); y = consumer(x); return z, y

**Pointwise consumers** — clear win at large shapes (21-28% faster):

| Consumer | `[1024,768]` | `[16384,768]` | `[30522,768]` |
|----------|-------------|--------------|--------------|
| add | 28.1 vs 26.5 µs (+5.8%) | 78.7 vs 103.6 µs (**-24.0%**) | 141.7 vs 180.3 µs (**-21.4%**) |
| mul | 28.0 vs 26.9 µs (+4.3%) | 78.9 vs 103.4 µs (**-23.7%**) | 141.2 vs 180.3 µs (**-21.7%**) |
| relu | 27.7 vs 26.3 µs (+5.1%) | 78.9 vs 103.5 µs (**-23.8%**) | 141.6 vs 180.4 µs (**-21.5%**) |
| to_fp16 | 28.2 vs 26.5 µs (+6.3%) | 66.2 vs 91.5 µs (**-27.7%**) | 116.8 vs 158.5 µs (**-26.3%**) |
| sin | 27.8 vs 26.5 µs (+5.2%) | 79.0 vs 102.8 µs (**-23.2%**) | 141.7 vs 181.7 µs (**-22.0%**) |

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos @shunting314 @eellison @BoyuanFeng

### CODE REVIEW

File: test/inductor/test_pad_as_cat.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        ref = fn(x)
+
+        self.assertEqual(result[0], ref[0])
+        self.assertEqual(result[1], ref[1])
+        kernel_count = _count_triton_kernels(code)
+        self.assertLessEqual(
+            kernel_count,
+            2,
+            f"Expected at most 2 kernels (fused + fill), got {kernel_count}.",
+        )
```
Comment (shunting314) [inline test/inductor/test_pad_as_cat.py]: The test will pass with or w/o the PR, right?

Maybe check num_bytes_accessed (https://github.com/pytorch/pytorch/blob/a12a78b0f4e64c23ce0083925c118dd138c14c00/torch/_inductor/metrics.py#L27)

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -2023,10 +2023,28 @@ def additional_pointwise_ops(op: torch._ops.OpOverload):
...
+            if not isinstance(fx_args, (list, tuple)):
+                return False
+            return any(hasattr(arg, "users") and len(arg.users) > 1 for arg in fx_args)
+
+        has_multi_consumers = any_input_has_multi_consumers()
+
+        horizontal_fuse_cat = (
+            all(should_lower_cat_input(inp) for inp in inputs)
+            and not any(can_fuse_reduction(t) for t in inputs)
+            and not has_multi_consumers
```
Comment (shunting314) [inline torch/_inductor/lowering.py]: `and not has_multi_consumers` part is redundant here

File: torch/_inductor/lowering.py [hunk 3]
```diff
@@ -2023,10 +2023,28 @@ def additional_pointwise_ops(op: torch._ops.OpOverload):
...
+        # Skip pointwise_cat when any input has multiple consumers. ConcatKernel
+        # NonOwningLayout lets all consumers share one realized copy.
+        def any_input_has_multi_consumers() -> bool:
+            cat_node = V.current_node
+            if cat_node is None:
+                return False
+            fx_args = cat_node.args[0]  # aten.cat format: cat(input_list, dim)
+            if not isinstance(fx_args, (list, tuple)):
+                return False
+            return any(hasattr(arg, "users") and len(arg.users) > 1 for arg in fx_args)
```
Comment (shunting314) [inline torch/_inductor/lowering.py]: Do you also want to check the user is fusible.

E.g. what if the user is a matmul

File: torch/_inductor/lowering.py (hunk 2 above)
Comment (tianrengao) [reply torch/_inductor/lowering.py]: removed the redundancy

File: torch/_inductor/lowering.py (hunk 3 above)
Comment (tianrengao) [reply torch/_inductor/lowering.py]: Added a check so Non-fusible users such as matmul will still use pointwise_cat

File: test/inductor/test_pad_as_cat.py (hunk 1 above)
Comment (tianrengao) [reply test/inductor/test_pad_as_cat.py]: updated tests to check num_bytes_accessed

File: torch/_inductor/lowering.py [hunk 4]
```diff
@@ -2023,10 +2024,36 @@ def additional_pointwise_ops(op: torch._ops.OpOverload):
...
+                return False
+            fx_args = cat_node.args[0]  # aten.cat format: cat(input_list, dim)
+            if not isinstance(fx_args, (list, tuple)):
+                return False
+
+            def has_fusible_multi_consumer(arg):
+                if not hasattr(arg, "users") or len(arg.users) <= 1:
+                    return False
+                return any(
+                    is_pointwise_use(u) for u in arg.users if u is not cat_node
```
Comment (shunting314) [inline torch/_inductor/lowering.py]: reduction should also be fine?

**Review by shunting314 — APPROVED** [review]

File: test/inductor/test_max_autotune.py [hunk 5]
```diff
@@ -1016,8 +1016,12 @@ def f(x, y):
...
         _, code = run_and_get_code(f_c, inps[0], inps[1])
         self.assertEqual(f_c(*inps), f(*inps), atol=0.03, rtol=0.25)
+        # x has multi-consumers (cat + add), so cat skips pointwise_cat
+        # and uses ConcatKernel.  cos is no longer fused into cat, giving
+        # one extra kernel launch compared to the single-consumer case.
+        count = 3 if (using_triton_mm or config.triton.native_matmul) else 2
         FileCheck().check(get_func_call()).check_count(
-            get_kernel_launch(), 2, exactly=True
+            get_kernel_launch(), count, exactly=True
         ).run(code[0])
```
Comment (tianrengao) [inline test/inductor/test_max_autotune.py:1025]: @shunting314 Our optimization may change cat multi-consumers like this unit test. 
```python
            y = torch.cos(y)
            x = torch.mm(x, x)
            out = torch.cat([x, y])
            return out, x + 1
```
before: kernel1 mm, kernel2 cos + cat + add. (mm read twice)
after: kernel1 mm, kernel2 cos + cat, kernel3 add. (mm read once, but `add` becomes a single kernel)

For memory bounded kernels(eg. large tensors in vllm cases), after is better. But for computing bounded kernels(small tensors), maybe this optimization should not apply. Do you think this is acceptable?

To fix this, a short term solution is to set a threshold for small tensors to go with pointwise_cat.
A longer term solution is adding a partial pointwise_cat: Instead of current all-or-nothing (all inputs inlined vs all concatkernel zero-copied), decide per input. Single consumer inputs get inlined into the cat kernel (preserving fusion). Multi-consumer inputs use NonOwningLayout zero-copy (avoiding redundant reads). This keeps kernel best of both pointwise_cat and ConcatKernel. This require some refactoring

File: torch/_inductor/lowering.py (hunk 4 above)
Comment (tianrengao) [reply torch/_inductor/lowering.py]: TL,DR: I just benchmarked concatKernel (with multi-consumer skip) vs pointwise_cat for pointwise and reduction. The conclusion is reduction benefits from pointwise but not concatkernel. So I will not skip pointwise_cat for reduction for now and I will land this pr first. Can revisit reduction if scheduler gains the ability to fuse Pointwise + Reduction nodes.


Test pattern: z = cat([x, zeros([6,768])]); y = consumer(x); return z, y

**Pointwise consumers** — clear win at large shapes (21-28% faster):

| Consumer | `[1024,768]` | `[16384,768]` | `[30522,768]` |
|----------|-------------|--------------|--------------|
| add | 28.1 vs 26.5 µs (+5.8%) | 78.7 vs 103.6 µs (**-24.0%**) | 141.7 vs 180.3 µs (**-21.4%**) |
| mul | 28.0 vs 26.9 µs (+4.3%) | 78.9 vs 103.4 µs (**-23.7%**) | 141.2 vs 180.3 µs (**-21.7%**) |
| relu | 27.7 vs 26.3 µs (+5.1%) | 78.9 vs 103.5 µs (**-23.8%**) | 141.6 vs 180.4 µs (**-21.5%**) |
| to_fp16 | 28.2 vs 26.5 µs (+6.3%) | 66.2 vs 91.5 µs (**-27.7%**) | 116.8 vs 158.5 µs (**-26.3%**) |
| sin | 27.8 vs 26.5 µs (+5.2%) | 79.0 vs 102.8 µs (**-23.2%**) | 141.7 vs 181.7 µs (**-22.0%**) |

**Reduction consumers** — no benefit at any shape:

| Consumer | `[1024,768]` | `[16384,768]` | `[30522,768]` | `[65536,768]` | `[131072,768]` |
|----------|-------------|--------------|--------------|--------------|---------------|
| sum | 41.0 vs 34.7 µs (+18.1%) | 79.1 vs 76.6 µs (+3.2%) | 142.1 vs 140.5 µs (+1.1%) | 304.2 vs 301.6 µs (+0.9%) | 576.6 vs 574.9 µs (+0.3%) |
| mean | 39.0 vs 32.5 µs (+19.8%) | 78.3 vs 76.7 µs (+2.1%) | 143.2 vs 140.6 µs (+1.8%) | 303.3 vs 301.6 µs (+0.6%) | 578.6 vs 575.3 µs (+0.6%) |
| amax | 39.3 vs 34.5 µs (+13.8%) | 79.6 vs 77.2 µs (+3.1%) | 143.1 vs 141.2 µs (+1.4%) | 303.4 vs 300.6 µs (+0.9%) | 574.8 vs 573.9 µs (+0.2%) |

Format: `ConcatKernel vs pointwise_cat (diff%)`. Negative = ConcatKernel faster.

I think the root cause of the difference is scheduler does not fuse reduction node and cat zero_fill.  

When ConcatKernel splits the cat, the zeros-fill becomes a separate IR node with a different iteration domain ([6,768]) from the main output ([N,768]). For pointwise consumers, the scheduler fuses zeros-fill with the consumer kernel. For reduction consumers, the scheduler cannot fuse a Pointwise node with a Reduction node. The extra kernel launch slows down it
